### PR TITLE
[C3] Improve bindings support in Svelte template 

### PR DESCRIPTION
**Pull request by jculvey**
**What this PR solves / how to test:**

- Updates the Svelte template add a hook to populate the `platform` object with `getPlatformProxy` in dev.
- Adds a default `wrangler.toml` file
- Adds a `build-cf-types` script to generate bindings types via `wrangler types`. 
- Updates the `app.d.ts` to include types for `ctx` and `cf` in addition to the `env` type described above
- Improves the reliability of parsing a project's `tsconfig.json` when adding the `workers-types` entrypoint by switching the implementation to `jsonc-parser`

**Author has addressed the following:**

- Tests
  - [ x ] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ x ] Included
  - [ ] Not necessary because:
- Associated docs
  - [ x ] Issue(s)/PR(s): https://github.com/cloudflare/cloudflare-docs/issues/12897
  - [ ] Not necessary because:

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

### CODE REVIEW

File: packages/create-cloudflare/e2e-tests/frameworks.test.ts [hunk 1]
```diff
@@ -468,6 +478,9 @@ const verifyBuildScript = async (
 		[pm, "run", script],
 		{
 			cwd: projectPath,
+			env: {
+				NODE_ENV: "production",
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/e2e-tests/frameworks.test.ts:482]: ooc: why is this needed? 😮

File: packages/create-cloudflare/templates/svelte/c3.ts [hunk 2]
```diff
@@ -39,39 +45,83 @@ const configure = async (ctx: C3Context) => {
...
 	configure,
-	transformPackageJson: async () => ({
-		scripts: {
-			"pages:preview": `${npm} run build && wrangler pages dev ${await compatDateFlag()} .svelte-kit/cloudflare`,
-			"pages:deploy": `${npm} run build && wrangler pages deploy .svelte-kit/cloudflare`,
-		},
-	}),
+	transformPackageJson: async (original: PackageJson, ctx: C3Context) => {
+		let scripts: Record<string, string> = {
+			preview: `${npm} run build && wrangler pages dev ${await compatDateFlag()} .svelte-kit/cloudflare`,
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/templates/svelte/c3.ts:110]: is the compat date necessary when we have it in the toml file?

(I sure hope not because I did remove it in the Next.js template 😅)

File: packages/create-cloudflare/src/templates.ts [hunk 3]
```diff
@@ -11,7 +11,12 @@ import degit from "degit";
 import { C3_DEFAULTS } from "helpers/cli";
 import { readJSON, usesTypescript, writeJSON } from "helpers/files";
 import { validateTemplateUrl } from "./validators";
-import type { C3Args, C3Context } from "types";
+import type { C3Args, C3Context, PackageJson } from "types";
+
+export type PackageTransformer = (
+	pkgJson: PackageJson,
+	ctx: C3Context
+) => Promise<Record<string, string | object>>;
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/src/templates.ts:19]: is this used anywhere?

Screenshot 2024-02-15 at 23 02 29

File: packages/create-cloudflare/templates/svelte/c3.ts [hunk 4]
```diff
@@ -39,39 +45,83 @@ const configure = async (ctx: C3Context) => {
...
+			deploy: `${npm} run build && wrangler pages deploy .svelte-kit/cloudflare`,
+		};
+
+		if (usesTypescript(ctx)) {
+			scripts = {
+				...scripts,
+				"build-cf-types": "wrangler types && mv worker-configuration.d.ts src/",
```
[the remark this answers is not in the document]
Comment (jculvey) [reply packages/create-cloudflare/templates/svelte/c3.ts:117]: It is, unfortunately adding this file to `include` in `tsconfig.json` doesn't work since the config inherits, and adding `include` overwrites rather than merges with inherited settings. I can't think of another way around this until `wrangler types` gets the enhancement we talked about.

File: packages/create-cloudflare/e2e-tests/frameworks.test.ts (hunk 1 above)
Comment (jculvey) [reply packages/create-cloudflare/e2e-tests/frameworks.test.ts:482]: This is needed because without it the `NODE_ENV` is set to `test`, which it inherits from the `vitest` call. When `svelte` is built with `NODE_ENV=test` it fails.

File: packages/create-cloudflare/templates/svelte/c3.ts (hunk 2 above)
Comment (jculvey) [reply packages/create-cloudflare/templates/svelte/c3.ts:110]: That's a great point 🙈

File: packages/create-cloudflare/src/templates.ts (hunk 3 above)
Comment (jculvey) [reply packages/create-cloudflare/src/templates.ts:19]: Good catch. It was gonna be, but it ended up being worse than what we have.

File: packages/create-cloudflare/templates/svelte/c3.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/create-cloudflare/templates/svelte/c3.ts:117]: isn't this going to be a problem on non unix machines?

File: packages/create-cloudflare/e2e-tests/frameworks.test.ts (hunk 1 above)
Comment (dario-piotrowicz) [reply packages/create-cloudflare/e2e-tests/frameworks.test.ts:482]: ah ok, thanks for the clarification 🙂

File: packages/create-cloudflare/templates/svelte/c3.ts (hunk 2 above)
Comment (dario-piotrowicz) [reply packages/create-cloudflare/templates/svelte/c3.ts]: PS: it might not be that simple thought 😓 

the compat date might not actually properly work with Pages: https://github.com/cloudflare/workers-sdk/issues/4411

I haven't had a chance I should investigate that issue 😓

File: packages/create-cloudflare/src/workers.ts [hunk 5]
```diff
@@ -105,32 +106,46 @@ export async function addWorkersTypesToTsConfig(ctx: C3Context) {
...
-				...(tsConfigJson.compilerOptions.types ?? []),
-				typesEntrypoint,
-			];
-			updated = JSON.stringify(tsConfigJson, null, 2);
-		} catch {
-			warn("Could not parse tsconfig.json file");
-			updated = tsconfig;
-		}
+		const explicitEntrypoint = (currentTypes as string[]).some((t) =>
+			t.match(/@cloudflare\/workers-types\/\d{4}-\d{2}-\d{2}/)
```
Comment (dario-piotrowicz) [inline packages/create-cloudflare/src/workers.ts:114]: Nit: should we also consider `@cloudflare/workers-types/experimental`?

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me 🙂 

Besides the failing test and `mv`
